### DISCUSSION TITLE
How to specify jmeter command line parameters?

### DISCUSSION THREAD

**Discussion by AntonioSun** [smicyk/groovy-jmeter#81] [Ideas]
<img width="514" alt="image" src="https://user-images.githubusercontent.com/2840074/172634582-534aa783-9a36-4511-ba66-6c7b76fca6f2.png">

the way I setup above variables allows them to be overridden when running jmeter on the command line.   
How to pass such parameters from the [groovy-jmeter](https://github.com/smicyk/groovy-jmeter) command line to jmeter's command line?

**Comment by AntonioSun** [comment]
Following up from #57,

My goal is total harmony between groovy-jmeter script and jmeter's. So, 

1. I want to specify variables in jmeter's way
1. and allow them to be overridden on jmeter's command line
1. while the source groovy-jmeter script works as well., without modification

But so far it is not working...

> OK. So some explanation about using variables.
> 
> You can use groovy variables in the script e.g.:
> 
> ```groovy
> def rampUpValue = 100
> 
> plan {
>     group users: 1, rampUp: rampUpValue, {
>        // other stuff
>     }
> }
> ```
> 
> So you can use variables define in the script and use them where you want in the script but they are used during building the script not during execution.
> 
> ```groovy
> plan {
>     variables {
>        variable name: 'rampUpValue', value, '100'
>     }
>     group users: 1, rampUp: '${rampUpValue}', {
>        // other stuff
>     }
> }
> ```
> 
> So the rampUpValue is not jmeter variable and is used during script execution not building. Note the single quotes.
> 
> So your example
> 
> ```groovy
> group(users: ${c_lt_users}, rampUp: ${c_lt_ramp}) {
> ```
> 
> Groovy doesn't know what is $ because it is not part of the language. But if you do:
> 
> ```groovy
> group(users: '${c_lt_users}', rampUp: '${c_lt_ramp}') {
> ```
> 
> It should work since you define c_lt_users as test plan variable, so jmeter would be able to evaluate it during script execution.


The problem with it is for point # 3 -- having the source groovy-jmeter script works with JMeter without modification.

Suppose ` group(users: '${c_lt_users}', rampUp: '${c_lt_ramp}') ` works for groovy-jmeter script, but it'll cause problem when converting that script into JMeter, when I'll get the error of

`Caught: java.lang.NumberFormatException: For input string: "${c_lt_users}"`

during the groovy-jmeter script to JMeter conversion.

If I don't use single quote, i.e., ` group(users: ${c_lt_users}, rampUp: ${c_lt_ramp})`, then groovy will fail, as you've just described.


_**In summary**_, the challenge is having the source groovy-jmeter script works with JMeter _without modification_.


> The example with jmx file which you send are fine if you didn't run them in JMeter GUI, the groovy script can be serialize to jmx with whatevery you put there but when your run it it will fail. The expressions are evaluated during runtime not during building jmx.

I don't quite grasp what you meant but as confirmed in https://github.com/smicyk/groovy-jmeter/issues/57#issuecomment-1149048226, a change like the following (which use the expressions in JMeter) works well.

```diff
@@ -76,3 +76,3 @@
         <stringProp name="ThreadGroup.on_sample_error">continue</stringProp>
-        <stringProp name="ThreadGroup.num_threads">1</stringProp>
+        <stringProp name="ThreadGroup.num_threads">${c_lt_users}</stringProp>
         <stringProp name="ThreadGroup.ramp_time">1</stringProp>
@@ -202,2 +202,7 @@
         </hashTree>
+        <UniformRandomTimer guiclass="UniformRandomTimerGui" testclass="UniformRandomTimer" testname="Uniform Random Timer" enabled="true">
+          <stringProp name="ConstantTimer.delay">${c_tt_delay}</stringProp>
+          <stringProp name="RandomTimer.range">${c_tt_range}</stringProp>
+        </UniformRandomTimer>
+        <hashTree/>
         <TransactionController guiclass="TransactionControllerGui" testclass="TransactionController" testname="TR0_SendMqMessage.php" enabled="true">
```

**Comment by smicyk** [comment]
Yes, I see now that I didn't look close enough to exception information and in case of groovy script the information comes from one of the validators. In my GUI test the error appears because I didn't initialize the variable.

Anyway, here we have two things to do:
1. add possibility to define (override) jmeter properties on groovy script command line

```shell
// for jmeter execution
jmeter -n -t script.jmx -l script.jtl -Jc_lt_users=2 -Jc_lt_ramp=5 -Jc_lt_range=6000 -Jc_lt_delay=2000
// for groovy execution
groovy script.groovy -Jc_lt_users=2 -Jc_lt_ramp=5 -Jc_lt_range=6000 -Jc_lt_delay=2000
```
So the groovy script will interpret -J as properties for jmeter script

2. Check if value for property in the scripts is a expression (contains ${}), if yes it should not do validation for that field but just pass it to the jmeter.

Both seems to be possible to implement.

**Reply by smicyk** [reply]
The fix for both things is in #84

**Reply by AntonioSun** [reply]
🎉🎉🎉! Looking forward to the new release!

**Reply by smicyk** [reply]
The new release is available https://github.com/smicyk/groovy-jmeter/releases/tag/v0.18.0.
